### DISCUSSION TITLE
Very long-term suggestion: add Lua as espanso's fully-fledged internal script language

### DISCUSSION THREAD

**Discussion by tmsg-gh** [espanso/espanso#1237] [Feature requests and Ideas]
Lua is a small (~200kb compiled) but powerful scripting language. It is written in pure ANSI C and v portable. There are already Rust bindings, see for instance https://docs.rs/crate/rlua/0.17.1 .

I have used Lua in two different scenarios to add scripting to an application and in both cases it was an easy and straightforward process.

This is not a suggestion for now or even for the next few months. It's rather an idea for a long-term addition to espanso.

**Comment by JeppeKlitgaard** [comment]
See also tangentially related issue: #679

Lua could indeed be a more powerful option if @federico-terzi  wants to go down the route of an outright scripting language.

It should be noted that there are also some fairly slim JS engines available for Rust. These could also be good options, though may be overkill in this application. The advantage would be that more people would know JS from elsewhere. That being said, as I recall it Lua is very quick to learn.

**Comment by tmsg-gh** [comment]
Weeeeeell, JS... I do think (with my Computer Science hat in place) that this is an abomination though I accept that it's an unavoidable abomination if you need to execute web client code (I use it myself for my personal website). Lua is a lot more straightforward and orthogonal than JS (of course IMO and I am perhaps ever so slightly biased).

One advantage I see for Lua is that in my experience people new to this whole programming thing take to it relatively easily. Case in point: Lua almost completely gets rid of those pesky strings of special characters that make Perl code such a joy to read though that comes at the price of having to write more code that looks a bit more longwinded to some... but especially beginners seem quicker to come to grips with that sort of almost "natural language" code. Again IMO.

**Comment by federico-terzi** [comment]
@tmsg-gh @JeppeKlitgaard Thank you for the feedback!

I've personally used Lua before and I agree it's very well-fit to be the scripting language for use-cases like this. That said, my biggest worry is always related to barriers to entry. 

Despite being an abomination under multiple point-of-views, most people know JS already and therefore could start using Espanso's Macros right away. The same _can't_ be said about Lua, as it's not as popular. Moreover, those that don't know JS would have a similar learning curve, as Lua shouldn't be significantly easier to learn.

I'm not excluding Lua support entirely, but the main deciding factor will be the complexity of the JS integration. If it turns out to be too complex, Lua might have a better chance :)

**Comment by tmsg-gh** [comment]
> That said, my biggest worry is always related to barriers to entry.

Exactly :-/ and I do think that Lua has a lower barrier than JS for those that are new to scripting. I honestly couldn't care less whether espanso supports JS or Lua, as I am fluent in both. But a newb?!

> most people know JS already and therefore could start using Espanso's Macros right away.

Most people? Well, most people I know who are using PCs but aren't programmers don't know **any** programming language at all.

> Moreover, those that don't know JS would have a similar learning curve, as Lua shouldn't be significantly easier to learn.

I beg to differ. I think Lua is significantly easier to learn than JS for a complete newbie. All this IMO :smile:

**Comment by federico-terzi** [comment]
@tmsg-gh Thank you for sharing your thoughts! Your points are definitely valid, I'll think about it. In the end, the question to answer would be: _do we want to optimize for newbies or programmers?_ And I can't answer that question right now :)

**Comment by tmsg-gh** [comment]
> do we want to optimize for newbies or programmers?

Yep, that's it in a nutshell. From reading other issues/threads I got a strong impression that your long-term goal is to move espanso beyond the "programmers' community" to all users. I may be mistaken though.

**Comment by archetyped** [comment]
Long-time JS user, recent Lua learner (Hammerspoon!), and I lean toward Lua as a preferred in-app scripting language as it seems like a better fit for needs outside of a web browser (standard libraries for IO, OS-specific values, etc.).  I also don't love the idea of adopting a language primarily because of its (current) popularity.

In terms of barrier to entry...

* Programmers: Should be able to pick up Lua very quickly (e.g. [Learn Lua in Y Minutes](https://learnxinyminutes.com/docs/lua/))
* Non-Programmers: About the same learning curve (Lua or JS) to get started and do simple things.

In the end, I'll be more than happy to see *any* fully-fledged language built into Espanso :)

**Comment by federico-terzi** [comment]
@archetyped thanks for sharing your thoughts!  Those points are definitely valid :)
